### PR TITLE
RFR: Cleanup

### LINKED ISSUE #22: Don't overwrite local modules in lit install
Also don't overwrite local modules in lit make

### LINKED ISSUE #23: Merge import logic in `lit add` and `lit make`

### PR DESCRIPTION
**Pull request by creationix**
This is a fairly major refactor of the inner guts to the client-side of lit.  The new structure is documented in `lib/README`.

This client does work with the existing production server.  This branch is ready for final testing.

Some important new features:

 - `lit make` and `lit install` don't ignore already installed modules in the modules folder.
 - `lit make` and `lit add` both ignore the `modules` folder when scanning for normal files.
 - `lit install` is faster for large projects.  It tries to do a lot more I/O in parallel so you don't have to wait for hundreds of TCP roundtrips.

When complete this closes #22, #23, #28, #30, and #31.  Depends on #29

### CODE REVIEW

**Comment by creationix** [comment]
The old client tests pass without any modification when running against the `test-server.lua` process from the master branch.

```
➜  lit git:(cleanup) ./tests/test-push.sh 
lit version: 0.2.1
command: auth creationix
load config: /Users/tim/lit/test-push/config
update config: /Users/tim/lit/test-push/config
username: creationix
name: Tim Caswell
email: tim@creationix.com
privateKey: /Users/tim/.ssh/id_rsa
checking ssh fingerprint: 8b:70:91:a9:39:02:68:c5:4b:b8:80:fe:b3:78:ec:3f
github request: https://api.github.com/users/creationix/keys
imported key: creationix c7:65:a1:8d:ef:b9:15:8f:80:20:e0:4f:19:27:60:7f
imported key: creationix fa:76:5c:0b:cb:da:0b:5f:b2:6b:1e:ba:d7:e0:ce:ee
imported key: creationix 8b:70:91:a9:39:02:68:c5:4b:b8:80:fe:b3:78:ec:3f
imported key: creationix 5e:45:80:db:e5:5c:a6:76:46:d0:04:d9:d0:95:22:14
imported key: creationix 59:33:10:ea:27:75:a4:0d:d8:e0:43:9b:da:d1:3e:1b
imported key: creationix 3c:26:40:1d:c2:ce:b1:f0:b3:3c:4f:62:43:02:81:ec
imported key: creationix 0a:56:d2:8a:c0:76:41:be:8d:21:f3:a8:e9:8f:ff:68
imported key: creationix 89:08:c2:48:49:f6:f8:dd:df:e3:fc:1c:a0:7f:3a:ec
imported key: creationix 3d:23:c5:6d:ed:3b:22:ab:76:99:62:55:91:05:29:e1
imported key: creationix a6:53:26:2d:92:04:1e:86:90:76:72:1e:cd:e4:d3:19
update config: /Users/tim/lit/test-push/config
done: success
lit version: 0.2.1
command: up ws://localhost:4822
load config: /Users/tim/lit/test-push/config
upstream: ws://localhost:4822
update config: /Users/tim/lit/test-push/config
done: success
lit version: 0.2.1
command: publish app/modules/readline.lua
load config: /Users/tim/lit/test-push/config
new tag: creationix/readline/v1.0.1
connecting: ws://localhost:4822
publishing: creationix/readline@v1.0.1
done: success
lit version: 0.2.1
command: publish app/modules/coro-channel.lua
load config: /Users/tim/lit/test-push/config
new tag: creationix/coro-channel/v1.0.1
connecting: ws://localhost:4822
publishing: creationix/coro-channel@v1.0.1
done: success
lit version: 0.2.1
command: publish app/modules/coro-fs.lua
load config: /Users/tim/lit/test-push/config
new tag: creationix/coro-fs/v1.2.3
connecting: ws://localhost:4822
publishing: creationix/coro-fs@v1.2.3
done: success
lit version: 0.2.1
command: publish app/modules/coro-tcp.lua
load config: /Users/tim/lit/test-push/config
new tag: creationix/coro-tcp/v1.0.0
connecting: ws://localhost:4822
publishing: creationix/coro-tcp@v1.0.0
done: success
lit version: 0.2.1
command: publish app/modules/coro-tls.lua
load config: /Users/tim/lit/test-push/config
new tag: creationix/coro-tls/v1.0.0
connecting: ws://localhost:4822
publishing: creationix/coro-tls@v1.0.0
done: success
lit version: 0.2.1
command: publish app/modules/git.lua
load config: /Users/tim/lit/test-push/config
new tag: creationix/git/v0.1.0
connecting: ws://localhost:4822
publishing: creationix/git@v0.1.0
done: success
lit version: 0.2.1
command: publish app/modules/hex-bin.lua
load config: /Users/tim/lit/test-push/config
new tag: creationix/hex-bin/v1.0.0
connecting: ws://localhost:4822
publishing: creationix/hex-bin@v1.0.0
done: success
lit version: 0.2.1
command: publish app/modules/http-codec.lua
load config: /Users/tim/lit/test-push/config
new tag: creationix/http-codec/v0.1.4
connecting: ws://localhost:4822
publishing: creationix/http-codec@v0.1.4
done: success
lit version: 0.2.1
command: publish app/modules/json.lua
load config: /Users/tim/lit/test-push/config
new tag: creationix/json/v2.5.0
connecting: ws://localhost:4822
publishing: creationix/json@v2.5.0
done: success
lit version: 0.2.1
command: publish app/modules/pretty-print
load config: /Users/tim/lit/test-push/config
new tag: creationix/pretty-print/v0.1.0
connecting: ws://localhost:4822
publishing: creationix/pretty-print@v0.1.0
done: success
lit version: 0.2.1
command: publish app/modules/prompt.lua
load config: /Users/tim/lit/test-push/config
new tag: creationix/prompt/v1.0.0
connecting: ws://localhost:4822
publishing: creationix/prompt@v1.0.0
done: success
lit version: 0.2.1
command: publish app/modules/require.lua
load config: /Users/tim/lit/test-push/config
new tag: creationix/require/v1.0.2
connecting: ws://localhost:4822
publishing: creationix/require@v1.0.2
done: success
lit version: 0.2.1
command: publish app/modules/semver.lua
load config: /Users/tim/lit/test-push/config
new tag: creationix/semver/v1.0.1
connecting: ws://localhost:4822
publishing: creationix/semver@v1.0.1
done: success
lit version: 0.2.1
command: publish app/modules/ssh-rsa.lua
load config: /Users/tim/lit/test-push/config
new tag: creationix/ssh-rsa/v0.1.2
connecting: ws://localhost:4822
publishing: creationix/ssh-rsa@v0.1.2
done: success
lit version: 0.2.1
command: publish app/modules/test-lib
load config: /Users/tim/lit/test-push/config
new tag: creationix/test-lib/v0.0.1
connecting: ws://localhost:4822
publishing: creationix/test-lib@v0.0.1
done: success
lit version: 0.2.1
command: publish app/modules/websocket-codec.lua
load config: /Users/tim/lit/test-push/config
new tag: creationix/websocket-codec/v1.0.0
connecting: ws://localhost:4822
publishing: creationix/websocket-codec@v1.0.0
done: success
➜  lit git:(cleanup) ./tests/test-pull.sh 
lit version: 0.2.1
command: auth creationix
load config: /Users/tim/lit/test-pull/config
update config: /Users/tim/lit/test-pull/config
username: creationix
name: Tim Caswell
email: tim@creationix.com
privateKey: /Users/tim/.ssh/id_rsa
checking ssh fingerprint: 8b:70:91:a9:39:02:68:c5:4b:b8:80:fe:b3:78:ec:3f
github request: https://api.github.com/users/creationix/keys
imported key: creationix c7:65:a1:8d:ef:b9:15:8f:80:20:e0:4f:19:27:60:7f
imported key: creationix fa:76:5c:0b:cb:da:0b:5f:b2:6b:1e:ba:d7:e0:ce:ee
imported key: creationix 8b:70:91:a9:39:02:68:c5:4b:b8:80:fe:b3:78:ec:3f
imported key: creationix 5e:45:80:db:e5:5c:a6:76:46:d0:04:d9:d0:95:22:14
imported key: creationix 59:33:10:ea:27:75:a4:0d:d8:e0:43:9b:da:d1:3e:1b
imported key: creationix 3c:26:40:1d:c2:ce:b1:f0:b3:3c:4f:62:43:02:81:ec
imported key: creationix 0a:56:d2:8a:c0:76:41:be:8d:21:f3:a8:e9:8f:ff:68
imported key: creationix 89:08:c2:48:49:f6:f8:dd:df:e3:fc:1c:a0:7f:3a:ec
imported key: creationix 3d:23:c5:6d:ed:3b:22:ab:76:99:62:55:91:05:29:e1
imported key: creationix a6:53:26:2d:92:04:1e:86:90:76:72:1e:cd:e4:d3:19
update config: /Users/tim/lit/test-pull/config
done: success
lit version: 0.2.1
command: up ws://localhost:4822
load config: /Users/tim/lit/test-pull/config
upstream: ws://localhost:4822
update config: /Users/tim/lit/test-pull/config
done: success
lit version: 0.2.1
command: install
load config: /Users/tim/lit/test-pull/config
connecting: ws://localhost:4822
matching: creationix/pretty-print 0.1.0
matching: creationix/require 1.0.2
matching: creationix/coro-fs 1.2.3
matching: creationix/coro-tcp 1.0.0
matching: creationix/coro-tls 1.0.0
matching: creationix/hex-bin 1.0.0
matching: creationix/semver 1.0.1
matching: creationix/git 0.1.0
matching: creationix/prompt 1.0.0
matching: creationix/ssh-rsa 0.1.2
matching: creationix/http-codec 0.1.4
matching: creationix/websocket-codec 1.0.0
matching: creationix/json 2.5.0
fetching: 13 objects
fetching: 13 objects
fetching: 3 objects
connecting: ws://localhost:4822
matching: creationix/coro-channel 1.0.0
fetching: 1 object
fetching: 1 object
matching: creationix/readline 1.0.1
fetching: 1 object
fetching: 1 object
matching: creationix/hex-bin 1.0.0
installing: creationix/ssh-rsa@0.1.2 -> modules/ssh-rsa.lua
installing: creationix/pretty-print@0.1.0 -> modules/pretty-print/
installing: creationix/coro-tls@1.0.0 -> modules/coro-tls.lua
installing: creationix/coro-tcp@1.0.0 -> modules/coro-tcp.lua
installing: creationix/hex-bin@1.0.0 -> modules/hex-bin.lua
installing: creationix/require@1.0.2 -> modules/require.lua
installing: creationix/json@2.5.0 -> modules/json.lua
installing: creationix/semver@1.0.1 -> modules/semver.lua
installing: creationix/coro-fs@1.2.3 -> modules/coro-fs.lua
installing: creationix/coro-channel@1.0.1 -> modules/coro-channel.lua
installing: creationix/websocket-codec@1.0.0 -> modules/websocket-codec.lua
installing: creationix/readline@1.0.1 -> modules/readline.lua
installing: creationix/git@0.1.0 -> modules/git.lua
installing: creationix/http-codec@0.1.4 -> modules/http-codec.lua
installing: creationix/prompt@1.0.0 -> modules/prompt.lua
done: success
/Users/tim/lit
lit version: 0.2.1
command: install
load config: /Users/tim/lit/test-pull/config
connecting: ws://localhost:4822
matching: creationix/pretty-print 0.1.0
matching: creationix/require 1.0.2
matching: creationix/coro-fs 1.2.3
matching: creationix/coro-tcp 1.0.0
matching: creationix/coro-tls 1.0.0
matching: creationix/hex-bin 1.0.0
matching: creationix/semver 1.0.1
matching: creationix/git 0.1.0
matching: creationix/prompt 1.0.0
matching: creationix/ssh-rsa 0.1.2
matching: creationix/http-codec 0.1.4
matching: creationix/websocket-codec 1.0.0
matching: creationix/json 2.5.0
connecting: ws://localhost:4822
matching: creationix/coro-channel 1.0.0
matching: creationix/readline 1.0.1
matching: creationix/hex-bin 1.0.0
installing: creationix/ssh-rsa@0.1.2 -> modules/ssh-rsa.lua
installing: creationix/pretty-print@0.1.0 -> modules/pretty-print/
installing: creationix/coro-tls@1.0.0 -> modules/coro-tls.lua
installing: creationix/coro-tcp@1.0.0 -> modules/coro-tcp.lua
installing: creationix/hex-bin@1.0.0 -> modules/hex-bin.lua
installing: creationix/require@1.0.2 -> modules/require.lua
installing: creationix/json@2.5.0 -> modules/json.lua
installing: creationix/semver@1.0.1 -> modules/semver.lua
installing: creationix/coro-fs@1.2.3 -> modules/coro-fs.lua
installing: creationix/coro-channel@1.0.1 -> modules/coro-channel.lua
installing: creationix/websocket-codec@1.0.0 -> modules/websocket-codec.lua
installing: creationix/readline@1.0.1 -> modules/readline.lua
installing: creationix/git@0.1.0 -> modules/git.lua
installing: creationix/http-codec@0.1.4 -> modules/http-codec.lua
installing: creationix/prompt@1.0.0 -> modules/prompt.lua
done: success
/Users/tim/lit
lit version: 0.2.1
command: down
load config: /Users/tim/lit/test-pull/config
upstream: disabled
update config: /Users/tim/lit/test-pull/config
done: success
lit version: 0.2.1
command: install
load config: /Users/tim/lit/test-pull/config
installing: creationix/ssh-rsa@0.1.2 -> modules/ssh-rsa.lua
installing: creationix/pretty-print@0.1.0 -> modules/pretty-print/
installing: creationix/coro-tls@1.0.0 -> modules/coro-tls.lua
installing: creationix/coro-tcp@1.0.0 -> modules/coro-tcp.lua
installing: creationix/hex-bin@1.0.0 -> modules/hex-bin.lua
installing: creationix/require@1.0.2 -> modules/require.lua
installing: creationix/json@2.5.0 -> modules/json.lua
installing: creationix/semver@1.0.1 -> modules/semver.lua
installing: creationix/coro-fs@1.2.3 -> modules/coro-fs.lua
installing: creationix/coro-channel@1.0.1 -> modules/coro-channel.lua
installing: creationix/websocket-codec@1.0.0 -> modules/websocket-codec.lua
installing: creationix/readline@1.0.1 -> modules/readline.lua
installing: creationix/git@0.1.0 -> modules/git.lua
installing: creationix/http-codec@0.1.4 -> modules/http-codec.lua
installing: creationix/prompt@1.0.0 -> modules/prompt.lua
done: success
/Users/tim/lit
➜  lit git:(cleanup) ./tests/test-offline.sh 
database: /Users/tim/lit/test-offline/db.git
storage: git
lit version: 0.2.1
command: auth creationix
load config: /Users/tim/lit/test-offline/config
update config: /Users/tim/lit/test-offline/config
username: creationix
name: Tim Caswell
email: tim@creationix.com
privateKey: /Users/tim/.ssh/id_rsa
checking ssh fingerprint: 8b:70:91:a9:39:02:68:c5:4b:b8:80:fe:b3:78:ec:3f
github request: https://api.github.com/users/creationix/keys
imported key: creationix c7:65:a1:8d:ef:b9:15:8f:80:20:e0:4f:19:27:60:7f
imported key: creationix fa:76:5c:0b:cb:da:0b:5f:b2:6b:1e:ba:d7:e0:ce:ee
imported key: creationix 8b:70:91:a9:39:02:68:c5:4b:b8:80:fe:b3:78:ec:3f
imported key: creationix 5e:45:80:db:e5:5c:a6:76:46:d0:04:d9:d0:95:22:14
imported key: creationix 59:33:10:ea:27:75:a4:0d:d8:e0:43:9b:da:d1:3e:1b
imported key: creationix 3c:26:40:1d:c2:ce:b1:f0:b3:3c:4f:62:43:02:81:ec
imported key: creationix 0a:56:d2:8a:c0:76:41:be:8d:21:f3:a8:e9:8f:ff:68
imported key: creationix 89:08:c2:48:49:f6:f8:dd:df:e3:fc:1c:a0:7f:3a:ec
imported key: creationix 3d:23:c5:6d:ed:3b:22:ab:76:99:62:55:91:05:29:e1
imported key: creationix a6:53:26:2d:92:04:1e:86:90:76:72:1e:cd:e4:d3:19
update config: /Users/tim/lit/test-offline/config
done: success
lit version: 0.2.1
command: down
load config: /Users/tim/lit/test-offline/config
upstream: disabled
update config: /Users/tim/lit/test-offline/config
done: success
lit version: 0.2.1
command: add app/modules/coro-channel.lua
load config: /Users/tim/lit/test-offline/config
new tag: creationix/coro-channel/v1.0.1
done: success
lit version: 0.2.1
command: add app/modules/coro-fs.lua
load config: /Users/tim/lit/test-offline/config
new tag: creationix/coro-fs/v1.2.3
done: success
lit version: 0.2.1
command: add app/modules/coro-tcp.lua
load config: /Users/tim/lit/test-offline/config
new tag: creationix/coro-tcp/v1.0.0
done: success
lit version: 0.2.1
command: add app/modules/coro-tls.lua
load config: /Users/tim/lit/test-offline/config
new tag: creationix/coro-tls/v1.0.0
done: success
lit version: 0.2.1
command: add app/modules/git.lua
load config: /Users/tim/lit/test-offline/config
new tag: creationix/git/v0.1.0
done: success
lit version: 0.2.1
command: add app/modules/hex-bin.lua
load config: /Users/tim/lit/test-offline/config
new tag: creationix/hex-bin/v1.0.0
done: success
lit version: 0.2.1
command: add app/modules/http-codec.lua
load config: /Users/tim/lit/test-offline/config
new tag: creationix/http-codec/v0.1.4
done: success
lit version: 0.2.1
command: add app/modules/json.lua
load config: /Users/tim/lit/test-offline/config
new tag: creationix/json/v2.5.0
done: success
lit version: 0.2.1
command: add app/modules/pretty-print
load config: /Users/tim/lit/test-offline/config
new tag: creationix/pretty-print/v0.1.0
done: success
lit version: 0.2.1
command: add app/modules/prompt.lua
load config: /Users/tim/lit/test-offline/config
new tag: creationix/prompt/v1.0.0
done: success
lit version: 0.2.1
command: add app/modules/readline.lua
load config: /Users/tim/lit/test-offline/config
new tag: creationix/readline/v1.0.1
done: success
lit version: 0.2.1
command: add app/modules/require.lua
load config: /Users/tim/lit/test-offline/config
new tag: creationix/require/v1.0.2
done: success
lit version: 0.2.1
command: add app/modules/semver.lua
load config: /Users/tim/lit/test-offline/config
new tag: creationix/semver/v1.0.1
done: success
lit version: 0.2.1
command: add app/modules/ssh-rsa.lua
load config: /Users/tim/lit/test-offline/config
new tag: creationix/ssh-rsa/v0.1.2
done: success
lit version: 0.2.1
command: add app/modules/test-lib
load config: /Users/tim/lit/test-offline/config
new tag: creationix/test-lib/v0.0.1
done: success
lit version: 0.2.1
command: add app/modules/websocket-codec.lua
load config: /Users/tim/lit/test-offline/config
new tag: creationix/websocket-codec/v1.0.0
done: success
lit version: 0.2.1
command: install
load config: /Users/tim/lit/test-offline/config
installing: creationix/ssh-rsa@0.1.2 -> modules/ssh-rsa.lua
installing: creationix/pretty-print@0.1.0 -> modules/pretty-print/
installing: creationix/coro-tls@1.0.0 -> modules/coro-tls.lua
installing: creationix/coro-tcp@1.0.0 -> modules/coro-tcp.lua
installing: creationix/hex-bin@1.0.0 -> modules/hex-bin.lua
installing: creationix/require@1.0.2 -> modules/require.lua
installing: creationix/json@2.5.0 -> modules/json.lua
installing: creationix/semver@1.0.1 -> modules/semver.lua
installing: creationix/coro-fs@1.2.3 -> modules/coro-fs.lua
installing: creationix/coro-channel@1.0.1 -> modules/coro-channel.lua
installing: creationix/websocket-codec@1.0.0 -> modules/websocket-codec.lua
installing: creationix/readline@1.0.1 -> modules/readline.lua
installing: creationix/git@0.1.0 -> modules/git.lua
installing: creationix/http-codec@0.1.4 -> modules/http-codec.lua
installing: creationix/prompt@1.0.0 -> modules/prompt.lua
done: success
➜  lit git:(cleanup) 
```

**Comment by creationix** [comment]
The next step for the server side is #28.  After that fixing server side should be trivial and this can be merged after some testing.

File: app/lib/db.lua [hunk 1]
```diff
@@ -1,346 +1,315 @@
...
+        local name = stack[#stack]()
+        if name then
+          local path = stack[#stack - 1]
+          local newPath = path and path .. "/" .. name or name
+          stack[#stack + 1] = newPath
+          stack[#stack + 1] = storage.nodes(prefix .. newPath)
+          return newPath
+        end
+        stack[#stack] = nil
+        stack[#stack] = nil
```
Comment (rphillips) [inline app/lib/db.lua]: double set to nil?

File: app/lib/rdb.lua [hunk 2]
```diff
@@ -57,162 +51,150 @@ return function (storage, url)
...
-    return remote.readAs("reply")
-  end
-
-  function upstream.match(name, version)
-    remote.writeAs("match", version and (name .. " " .. version) or name)
-    local data = remote.readAs("reply")
-    if data then
-      return string.match(data, "^([^ ]+) (.*)$")
+  function db.push(hash)
+    connect();
```
Comment (rphillips) [inline app/lib/rdb.lua]: semicolon

File: app/lib/rdb.lua [hunk 3]
```diff
@@ -57,162 +51,150 @@ return function (storage, url)
...
+        for i = 1, #data do
+          remote.writeAs("send", db.load(data[i]))
+        end
+      elseif name == "done" then
+        return data
+      else
+        error("Expected more wants or done in reply to send to server")
+      end
     end
+    disconnect();
```
Comment (rphillips) [inline app/lib/rdb.lua]: semicolon

File: app/lib/rdb.lua [hunk 4]
```diff
@@ -57,162 +51,150 @@ return function (storage, url)
...
-  --[[
-  upstream.close()
-  ----------------
-
-  Called when the db is done with the connection.
-  ]]--
-  function upstream.close()
-    return socket:close()
+  function db.upquery(name, request)
+    connect();
```
Comment (rphillips) [inline app/lib/rdb.lua]: semicolon

File: app/lib/db.lua (hunk 1 above)
Comment (creationix) [inline app/lib/db.lua]: Yep, it pops two items from the stack.

File: app/lib/rdb.lua (hunk 2 above)
Comment (creationix) [inline app/lib/rdb.lua]: thanks.

**Comment by creationix** [comment]
Waiting on #28, #29, and #30

**Comment by creationix** [comment]
This is done and ready to merge.

**Comment by rphillips** [comment]
on OSX: ```find: illegal option -- t``` when I run ```make```

after the nit, +1... upgraded the virgo-base:luvi-up branch... just needs a good way to get lit installed within the makefile. https://github.com/virgo-agent-toolkit/virgo-base-agent/tree/luvi-up

File: lib/db.lua [hunk 5]
```diff
@@ -0,0 +1,373 @@
+--[[
+
```
Comment (rphillips) [inline lib/db.lua]: love the comment here.

File: lib/storage.lua [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+    end
+  end
+
+  local function iter(path, filter)
+    local iter, err = fs.scandir(path)
+    if not iter then
+      if err:match("^ENOENT:") then
+        return function() end
+      end
+      assert(iter, err)
```
Comment (rphillips) [inline lib/storage.lua]: iter should always be non-nil here, right?

Comment (creationix) [inline lib/storage.lua]: If there was an error trying to scandir, `iter` will be nil and `err` will be the error message.  ENOENT isn't the only possible error case for `uv_fs_scandir`.

**Pull request merged**